src/init.cpp

```diff
@@ -388,9 +388,9 @@ void Shutdown(NodeContext& node)
     if (node.validation_signals) {
         node.validation_signals->UnregisterAllValidationInterfaces();
     }
+    node.chainman.reset();
```

Comment:
In commit ef83e9569752b4273e6aed3124642dc057cab540 (shutdown: Tear down mempool after chainman)

Feel free to disregard as out of scope, but while addressing this, it might be nice to make reintroducing this dangling pointer impossible by making the various mempool pointers `shared_ptr`, e.g. https://github.com/davidgumberg/bitcoin/commit/75c839b8d77c207a0946294672b703d9f8eecbe1

Reply:
Thanks for taking a look :)

I think the problem here is that we are moving the mempool pointer around between different objects, while they don't really share ownership of it at any one time. It not having a value at some points in time does have a logical meaning in the current code. In general I don't like shared pointers for a number of reasons. In this instance I agree that it would solve the problem of not knowing when to destruct the mempool. I am a bit apprehensive to commit this change here though, because I feel like the current design with a unique pointer and raw pointers signalling optional pointer values was done on purpose over a shared pointer design.